### PR TITLE
Ensure post-featured-image block is in_the_loop() for BC with core and plugins, and to fix lazy-loading

### PR DESCRIPTION
**Pull request by felixarntz**
## What?
This PR is part of the solution from https://github.com/WordPress/wordpress-develop/pull/3560 (see that PR for additional context), which addresses https://core.trac.wordpress.org/ticket/56930. The problem here specifically also addresses #41783.

## Why?
Currently the logic to not lazy-load the first featured image or in-content image is not working for block themes. https://github.com/WordPress/wordpress-develop/pull/3560 is the full solution to that problem, however this PR here is necessary since the block itself comes from Gutenberg. We should therefore merge and port this to WP core's blocks soon (in time for WP 6.2).

## How?
WP core's logic to determine whether to lazy-load an image expects to run `in_the_loop()` when the main content is rendered. Since block themes by default may not enter the loop initially, this extra call here needs to be added, as currently the lack of it is a backward compatibility break causing the problem outlined above. This fix is already present in the `post-content` block, so it's simply brought over here as well.

## Testing Instructions
1. Test this as part of https://github.com/WordPress/wordpress-develop/pull/3560 as by itself here it is not going to solve the entire problem.
2. Use a block theme and view a post with a featured image in the frontend.
3. With WordPress `trunk`, the first featured image will have `loading="lazy"` on it, which is incorrect. With https://github.com/WordPress/wordpress-develop/pull/3560, it will not have the attribute as expected.

### CODE REVIEW

File: packages/block-library/src/post-featured-image/index.php
Comment (adamsilverstein) [inline packages/block-library/src/post-featured-image/index.php:25]: same question as the core API, why not just set the in_the_loop global directly (the_post might have unintended consequences?)

Comment (felixarntz) [reply packages/block-library/src/post-featured-image/index.php:25]: @adamsilverstein I think setting only the in_the_loop global would have other unintended consequences.

To be fair, this code is not new, I copied it as is from the `post-content` block where this is already used in production since 1.5 years (see https://github.com/WordPress/wordpress-develop/commit/585f62ad8c929af0100467ce2d51fed749a02b4f#diff-7adcebab0e7d7e6e897a5854fd02fc2421ea6acfc1ccc21c651fde4ef5c0a3caR47, this is when it was introduced in core). Given that nobody has raised an issue about it since, I felt it was reasonable enough to reuse the same approach here.

**Comment by felixarntz** [comment]
@Mamaduka I'd like to get this on your radar. Can we prioritize this PR for an upcoming Gutenberg release and then a merge to WP core `trunk`, to unblock https://core.trac.wordpress.org/ticket/56930 and https://github.com/WordPress/wordpress-develop/pull/3560, which are scheduled for WP 6.2?

**Comment by Mamaduka** [comment]
Thanks for the ping, @felixarntz!

This is on my to-do list, and I will try to test it as soon as possible thoroughly.

File: packages/block-library/src/post-featured-image/index.php
Comment (Mamaduka) [inline packages/block-library/src/post-featured-image/index.php:30]: Should we remove the `$post_ID` usage from the `context` now that we set the global post object? I think we did something similar in #37622.

**Review by Mamaduka — APPROVED** [review]
The changes test well for me, and considering we already have similar code in the Post Content block for BC, I think this is safe to land.

## Tests

* I confirmed that thumbnail previews work as before.
* The first featured image doesn't have the `loading="lazy"` attribute when testing with https://github.com/WordPress/wordpress-develop/pull/3560.

## A follow-up

We'll need a similar follow-up for the Cover block when it renders the featured image, but this doesn't have to be a blocker for https://github.com/WordPress/wordpress-develop/pull/3560.

**Review by aristath — APPROVED** [review]

File: packages/block-library/src/post-featured-image/index.php
Comment (felixarntz) [reply packages/block-library/src/post-featured-image/index.php:30]: @Mamaduka Oh you mean not pass `$post_ID` to these functions and instead rely on the default which would be the same current post anyway? That seems reasonable to me.

We could do this here and also further down on `get_the_post_thumbnail()` and `get_the_permalink()`.

Do you think I should add it to this PR or rather in a separate PR?

Comment (Mamaduka) [reply packages/block-library/src/post-featured-image/index.php:30]: > Oh you mean not pass $post_ID to these functions and instead rely on the default which would be the same current post anyway? That seems reasonable to me.
> We could do this here and also further down on get_the_post_thumbnail() and get_the_permalink().

Correct.

> Do you think I should add it to this PR or rather in a separate PR?

We can do it in a separate PR if you prefer.

Comment (felixarntz) [reply packages/block-library/src/post-featured-image/index.php:30]: @Mamaduka Sounds good, I'll open a follow up PR once this is merged. Is this waiting for anything else or should we merge it?

Comment (Mamaduka) [reply packages/block-library/src/post-featured-image/index.php:30]: Feel free to merge 👍

Comment (felixarntz) [reply packages/block-library/src/post-featured-image/index.php:30]: I've opened #45828 to kickstart that work. Could use some guidance there on how to proceed.

Now that this is merged, will it automatically make it into WP 6.2? Or do we need to add it to a certain Gutenberg branch?

Comment (Mamaduka) [reply packages/block-library/src/post-featured-image/index.php:30]: Correct, this will be automatically included in WP 6.2. Usually, all non-experimental changes before the feature freeze are automatically included. This is when we create a dist branch - `wp/6.2`.

We only sync packages with the core repo before releases. Would that be a problem for landing https://github.com/WordPress/wordpress-develop/pull/3560?

Comment (felixarntz) [reply packages/block-library/src/post-featured-image/index.php:30]: Yeah, that's the main reason I'm asking. Is there any way we can cherry-pick / backport this change to core sooner? Happy to help as much as I can.

Comment (Mamaduka) [reply packages/block-library/src/post-featured-image/index.php:30]: We'll have to do a minor release for the packages outside the release scope, but I don't know what's the policy here. I'm going to ping folks who have done more release than me.

Cc. @noisysocks, @youknowriad

Comment (youknowriad) [reply packages/block-library/src/post-featured-image/index.php:30]: > We'll have to do a minor release for the packages outside the release scope, but I don't know what's the policy here. I'm going to ping folks who have done more release than me.

This is something we do often I think. We'll have to cherry-pick the commits to the right `wp/*` branch and perform a npm patch release from there. I'm not entirely certain with the process (as it changed lately a bit) but I know that it's documented.

Comment (Mamaduka) [reply packages/block-library/src/post-featured-image/index.php:30]: Thanks, Riad! I'm familiar with the new process and can do a patch release.

@felixarntz, I can do a package release and prepare core PR early next week.

Comment (felixarntz) [reply packages/block-library/src/post-featured-image/index.php:30]: Perfect, thanks both! Let me know if I can help with anything like reviews or so.

**Comment by tomusborne** [comment]
@felixarntz Just a heads up that this PR seems to have broken the core loop if the Post Featured Image block is hooked before it: https://core.trac.wordpress.org/ticket/58027

**Comment by felixarntz** [comment]
@tomusborne Thank you, looking into it!

**Pull request merged**
